Issue #2616: Add CatchParameterName check for catch blocks parameters

mkordas opened the pull request:
I've set the default pattern to the very strict one: `^(e|ex|[a-z][a-z][a-zA-Z]*)$`. I'm aware, that most probably it's too controversial, but I'm waiting for your opinions @romani @rnveach @Vladlis @MEZk 

This is early version of the check - any input/review is welcome.

rnveach commented:
> ^(e|ex|[a-z][a-z][a-zA-Z]*)$

I feel unless we change the other checks for better naming standards, we should leave the default the same as ParameterNameCheck.

MEZk commented:
@mkordas 
There is no need to specify 'ex'.
```*``` means between zero and unlimited times, as many times as possible, giving back as needed.

^(e|[a-z][a-z][a-zA-Z]*)$ will be enough.
Check yourself here: https://regex101.com/

Vladlis commented:
1) Agree with @MEZk, the proposed pattern is not correct.
2) Shouldn't the violations in checkstyle code be fixed in a separate commit?

mkordas commented:
@rnveach 
> I feel unless we change the other checks for better naming standards, we should leave the default the same as ParameterNameCheck.

ParameterName's pattern would prohibit `Exception e` and I have strong feeling that it should be allowed. If anything, we could take pattern from LocalVariableName: `^[a-z][a-zA-Z0-9]*$` - what do you think?

mkordas commented:
@Vladlis 
> Shouldn't the violations in checkstyle code be fixed in a separate commit?
Will do.

Vladlis commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheck.java:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * <p>
+ * Checks that <code>catch</code> parameter names conform to a format specified
```
{@code catch} can be used here

romani commented:
I love code review by 5 people :) .

More items from me:

1) As we completely separate validation of names by ParameterName and CatchParameterName. We need to add new Check to Google style guide, as all names should follow - https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names (no "$", no "_", two lowercase characters before first capital).
"e", "ex".... is ok by Google. Additional allowances are "ok", "expected" https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions .

2) Testing (html reports) on Guava should be provided. From this reports we need to catch other names for Exception objects. Other opensoure projects from Google should be used in testing to fill format with all reasonable names for catch parameters.

mkordas commented:
I've applied all the comments and Google's and Checkstyle's patterns are tuned. I've noticed that also `catch(Throwable t)` is common idiom in Java world, so I've allowed it.

Currently Checkstyle's pattern requires `e`, `t`, `ex` or at least 3 chars, while Google's accept `e`, `t` or anything that is at least 2 chars. Numbers are allowed respectively from third or second position.

Results of default config over Guava, Spring and Hibernate: http://mkordas.github.io/catch-parameter-name/checkstyle.html
And the same for Google Style config: http://mkordas.github.io/catch-parameter-name-google/checkstyle.html

Please review.

MEZk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/CatchParameterNameCheck.java:
```diff
@@ -0,0 +1,78 @@
+ * <p>
+ * Checks that {@code catch} parameter names conform to a format specified by the format property.
+ * The format is a {@link java.util.regex.Pattern regular expression} and defaults to
+ * <strong>^(e|ex|[a-z][a-z][a-zA-Z]*)$</strong>.
```
Default is "^(e|t|ex|[a-z][a-z][a-zA-Z0-9]+)$".

mkordas replied:
Thanks for spotting, I will fix it.

MEZk commented:
@mkordas 
Reports look good.
@romani 
Should we generate reports against Guava, since the check is for Google Style guide?

mkordas commented:
There is one more change to be done - update docs for Google Style Guide: http://checkstyle.sourceforge.net/google_style.html#a5.2.1

But where should I put this check? There is no dedicated section for in... Maybe it should be in `5.1 Rules common to all identifiers`?

mkordas commented:
* Updated default pattern
* Updated documentation:
![image](https://cloud.githubusercontent.com/assets/5467276/11283817/cd925a14-8f06-11e5-866e-9e7be081973d.png)
* Updated Google Style docs:
![image](https://cloud.githubusercontent.com/assets/5467276/11283889/2facd4c2-8f07-11e5-97a0-0811a1a30265.png)

>Should we generate reports against Guava with maven plugin, since the check is for Google Style guide?

@MEZk  what do you mean by that? I already provided such report: http://mkordas.github.io/catch-parameter-name-google/checkstyle.html

MEZk commented:
@mkordas 
I was confused that you did not exclude 'test' folders). Nevertheless, it is ok, since the check does not rise any violations on Guava except the one in your report.

mkordas commented:
I will fix the build today in the evening. In the meantime - am I going in right direction?

mkordas commented:
Build is fixed, @romani please do final code review.

romani commented:
we should prohibit numbering in names by default

Some statistics from your report:
se	56 - SecurityException + SQLException (project specific type )
he	32 - HibernateException (specific)
ok	22 - Guava only
ee	19 - just some unclear usage
me	15 - MappingException (specific)
pe	11 - ParseException (specific)
ie	10 - IOException
ce	5
e1	3
ne	3
qe	3
re	3
e	2
e2	2
es	2
nf	2
ae	1
ue	1
x	1

So I do not think it required to extend default format.

mkordas commented:
@romani - default pattern has been changed

romani commented:
@mkordas , your report is without links to code, please do not do this in future.

FYI: https://github.com/checkstyle/contribution/issues/40

mkordas commented:
I know, but xref generation and upload was killing my PC. checkstyle/contribution#40 will definitely help.

romani commented:
@mkordas , https://github.com/checkstyle/contribution/issues/40 is fixed, please regenerate report and confirm that all violations are valid.

mkordas commented:
@romani updated report: http://mkordas.github.io/catch-parameter-name-refined/checkstyle.html. I don't see any false-positives.

romani commented:
@mkordas , sorry for late reply.

1)
> ```<module name="CatchParameterName"/>```

please forbid single character "e" usage in Checkstyle code. User should use at-least "ex". Non-default config is good example for other users.

2)
> allows {@code ok} abbreviation


please remove "ok" from default, it is google's style not a general style.

3)
> google_checks.xml
name="format" value="^(e|t|[a-z][a-z0-9][a-zA-Z0-9]*)$"

https://google.github.io/styleguide/javaguide.html#s6.2-caught-exceptions
"ok" should be there, and I think we should add "expected" too.

4)
> google_style.xml
"5.2 Rules of identifier type" already includes this rule.

please do not remove this text. Please move check reference to "6.2 Caught exceptions: not ignored" and "5.2.6 Parameter names".


5)
> rule51identifiernames/InputCatchParameterName.java

in google's inputs we have "//violation" comment in a line that cause it, as config is strict - violations is always on the same place. It is not clear why you pass this validation easily, I thought we enforced these by code.

mkordas commented:
1) done
2) done, but why we want to restrict by default things that are allowed in Google Style and thus probably acceptable in majority of projects?
3) `ok` and `expected` are within `[a-z][a-z0-9][a-zA-Z0-9]*` pattern
4) "6.2 Caught exceptions: not ignored" doesn't say anything about naming convention. We also cannot assign this check to "5.2.6 Parameter names" because following what's written there we would need to prohibit one-letter catch parameters
5) I don't see any issue, can you clarify what is wrong?

mkordas commented:
@romani 
One more thought: I think that catch parameters are much closer to the local variables than method parameters. They have limited scope, they are never visible from outside and one letter names are widely accepted.

romani commented:
1)
>  but why we want to restrict by default things that are allowed in Google Style and thus probably acceptable in majority of projects?

from your report I did not see that ok is used at any other project, am I miss smth ?

2)
>  "6.2 Caught exceptions: not ignored" doesn't say anything about naming convention

Please search for "In tests, a caught exception may be ignored without comment if it is named expected", "} catch (NoSuchElementException expected) {" , "} catch (NumberFormatException ok) {"

3)
> "5.2.6 Parameter names" because following what's written there we would need to prohibit one-letter catch parameters

Yes and no. Google style is far from strict, we had a lot problems with it before. My intend to place it in two places to make me/you hint in future to answer "how checkstyle cover point 6.2 ?" - it covers by two Checks .... .

4)
> 5) I don't see any issue, can you clarify what is wrong?

I expect comments like :
https://github.com/mkordas/checkstyle/blob/914246dd2db0b4209642876727a319120005434f/src/it/resources/com/google/checkstyle/test/chapter5naming/rule522typenames/InputTypeName.java#L3

5)
> One more thought: I think that catch parameters are much closer to the local variables than method parameters.

But lets look at https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.20.1 , ` The value V is assigned to the parameter of the selected catch clause`. So java language think that is "parameter".
Syntax - https://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-Catches, `CatchClause:
catch ( CatchFormalParameter ) Block `. It is parameter.
We have to be strict in terminology, even most of software engineers use non-official terminology.

6)
> and one letter names are widely accepted.

not all letters, only "e", "t" and it only old-fasion habit that need to be eliminated. I can show you how Java team fix names from old-style to new (I saw some examples when I eliminated "URI" abbreviation that came as parameter from overridden method).

mkordas commented:
> from your report I did not see that ok is used at any other project, am I miss smth ? 

It is not used in Spring or Hibernate, but searching over GitHub shows a lot of occurences: https://github.com/search?q=%22catch+%28IllegalArgumentException+ok%29%22&type=Code
https://github.com/search?q=%22catch+%28NullPointerException+ok%29%22&type=Code
Still, I haven't found any other popular project like Guava where `ok` is used, so you may be right.

> Please search for "In tests, a caught exception may be ignored without comment if it is named expected", "} catch (NoSuchElementException expected) {" , "} catch (NumberFormatException ok) {"

I see that, but it doesn't say anything about naming convention. It only shows that names like `expected` or `ok` are allowed, but it's also clear from general rules for all identifiers.

> Yes and no. Google style is far from strict, we had a lot problems with it before.

Google defines [5.1 Rules common to all identifiers](https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names) so we may use just them as starting point.
>I expect comments like

You are right, fixed. Apparently the only verification that is happening is assertion that number of `// warn` comments is equal to actual warnings, please see https://github.com/mkordas/checkstyle/blob/914246dd2db0b4209642876727a319120005434f/src/it/java/com/google/checkstyle/test/base/BaseCheckTestSupport.java#L219. There is no logic for `ok` or `violation` comments.

>  CatchClause: catch ( CatchFormalParameter ) Block. It is parameter.

I would say that 'parameter' is something different than 'catch parameter' despite sharing the same name... but we may discuss it further.

> not all letters, only "e", "t" and it only old-fasion habit that need to be eliminated. 

Agreed. Anyway, I did some stats around [Java Corpus](https://github.com/tateology/java-corpus/) and it looks like also letter `x` is pretty popular.
```
   7673 catch (Exception e)
   5399 catch (IOException ex)
   4386 catch (Exception ex)
   4288 catch (IOException e)
   2116 catch (IOException ioe)
   1454 catch (InterruptedException ex)
    928 catch (InterruptedException e)
    873 catch (Throwable t)
    839 catch (BadLocationException ex)
    751 catch (IllegalArgumentException e)
    651 catch (Throwable e)
    651 catch (ClassNotFoundException e)
    595 catch (BadLocationException e)
    594 catch (MalformedURLException ex)
    591 catch (IllegalArgumentException ex)
    565 catch (NumberFormatException e)
    521 catch (RuntimeException rex)
    517 catch (Error err)
    487 catch (IllegalArgumentException expected)
    487 catch (ClassNotFoundException ex)
    475 catch (java.lang.Exception ex)
    463 catch (RuntimeException e)
    421 catch (DataObjectNotFoundException ex)
    369 catch (NumberFormatException nfe)
    356 catch (InvocationTargetException ex)
    348 catch (SVNClientException ex)
    322 catch (URISyntaxException ex)
    316 catch (IllegalAccessException ex)
    313 catch (BadLocationException ble)
    307 catch (IllegalAccessException e)
    304 catch (RecognitionException re)
    300 catch (IOException x)
    299 catch (NumberFormatException ex)
    295 catch (Exception x)
```

mkordas commented:
@romani, any feedback on the above? Travis failure is just some connectivity problem.

romani commented:
sorry for delay, I somehow miss your reply.

>  but it's also clear from general rules for all identifiers.

ok.

> Google defines 5.1 Rules common to all identifiers so we may use just them as starting point

ok, lets do this , but please add this Check also to https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names as it state `One-character parameter names should be avoided.`  and we for now treat it as parameter.

> but we may discuss it further

we need to stick to Java specification for terminology. I do not have more evidences. Using the official terminology is always good.

> There is no logic for ok or violation comments.

please change comment to "// warn" to be consistent with other, please create issue to make sure we will not lose such comments in future.

> looks like also letter x is pretty popular.

This is style of this project, I have never see "x" in my practice. It usage is rare in comparison to others, it might be unfortunate naming and there is no tool to point on this. I am against of update of default pattern.

1)
> google_checks.xml ```<property name="format" value="^(e|t|[a-z][a-z0-9][a-zA-Z0-9]*)$"/> ```

please remove "e|t" as https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names `One-character parameter names should be avoided.`

mkordas commented:
> This is style of this project, I have never see "x" in my practice. It usage is rare in comparison to others, it might be unfortunate naming and there is no tool to point on this. I am against of update of default pattern.

Agreed, I'm against `x` too.

> please remove "e|t" as https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names One-character parameter names should be avoided.

Are you really sure? `e` and `t` are used in many places in Guava: [catch (Exception e)](https://github.com/google/guava/search?p=2&q=%22catch+exception+e%22&type=Code&utf8=%E2%9C%93), [catch (Throwable t)](
https://github.com/google/guava/search?utf8=%E2%9C%93&q=%22catch+throwable+t%22&type=Code). My opinion is that the author of Google Style Guide by "parameters" meant only "method parameters", not "catch parameters". My intuition says that "catch parameters" should follow local variable naming conventions, as they are closest to them when it comes to impact of names to the surrounding code.

I will respond to the other points later.

romani commented:
> are used in many places in Guava

Guava violates Google style in 3K places, we can not treat this project as standard. We took Guava project for testing only because it was the only ease to find open-source project from Google at that time and following style is in demand for libraries as their code is read by millions of engineers. 

Style that is only written in paper and not enforced by tool will always be at status of dreams or Utopian novel (human factor). I do not know any engineer that could write a code that completely follow the style. Even I have fight with Checkstyle each time I touch a code.

> My opinion is that the author of Google Style Guide by "parameters" meant only "method parameters", not "catch parameters"

As we did for all other cases, you can see my issue against Guava https://github.com/google/guava/issues/1891 , we just follow to what is written in style guide, and wait for style update. No changes were done to Style guide even I pointed to the unclear parts of it and even Guava maintainers agree that style need to fixed.  Style guide need to fixed first, in other case we will have another part of java community complaining about style configuration and not following to what is written. Even guava people point me to trust to what is written few times , when we were not sure.

mkordas commented:
OK, you've convinced me. Thanks for fruitful discussion!
> please change comment to "// warn" to be consistent with other

done
> please create issue to make sure we will not lose such comments in future

#2658
>  please add this Check also to https://google.github.io/styleguide/javaguide.html#s5.2.6-parameter-names

done
> please remove "e|t"

done

romani commented:
merged as FF
